caffe2/core/plan_executor: add cancellation of async nets on error + propagate exceptions via std::exception_ptr for stack traces

d4l3k opened the pull request:
Summary:
This has two parts:
* When `--caffe2_handle_executor_threads_exceptions` is set when a parallel execution step throws an exception it can hang waiting for async nets to finish. This adds cancellation code to cancel any async nets.
* This makes the exceptions returned from parallel workers pass a std::exception_ptr so the stack trace can be recorded with folly::SmartExceptionTracer. This does use newer C++ features but per from recent discussion all C++14 features can be used now.

Test Plan: buck test //caffe2/caffe2:caffe2_test_cpu

Differential Revision: D19320177

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 268c8fa7:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=268c8fa7447564d0311878d401e05819909a7ffe).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4175210) pytorch_macos_10_13_py3_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=82229413) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=82229413))
<details>
<summary>
<code>Jan 08 14:21:33 FAILED: caffe2/CMakeFiles/plan_executor_test.dir/core/plan_executor_test.cc.o</code>
</summary>

```
Jan 08 14:21:31 ../aten/src/ATen/test/NamedTensor_test.cpp:53:22: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare] 
Jan 08 14:21:31   for (auto i = 0; i < names.size(); i++) { 
Jan 08 14:21:31                    ~ ^ ~~~~~~~~~~~~ 
Jan 08 14:21:31 1 warning generated. 
Jan 08 14:21:32 [1606/2988] Building CXX object caffe2/CMakeFiles/init_test.dir/core/init_test.cc.o 
Jan 08 14:21:32 [1607/2988] Building CXX object caffe2/CMakeFiles/scalar_tensor_test.dir/__/aten/src/ATen/test/scalar_tensor_test.cpp.o 
Jan 08 14:21:32 [1608/2988] Building CXX object caffe2/CMakeFiles/torch.dir/__/empty.cpp.o 
Jan 08 14:21:33 [1609/2988] Building CXX object caffe2/CMakeFiles/backend_fallback_test.dir/__/aten/src/ATen/test/backend_fallback_test.cpp.o 
Jan 08 14:21:33 [1610/2988] Building CXX object caffe2/CMakeFiles/kernel_lambda_test.dir/__/aten/src/ATen/core/boxing/kernel_lambda_test.cpp.o 
Jan 08 14:21:33 [1611/2988] Building CXX object caffe2/CMakeFiles/plan_executor_test.dir/core/plan_executor_test.cc.o 
Jan 08 14:21:33 FAILED: caffe2/CMakeFiles/plan_executor_test.dir/core/plan_executor_test.cc.o  
_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces "-Xpreprocessor -fopenmp -I/usr/local/include" -std=gnu++14 -MD -MT caffe2/CMakeFiles/plan_executor_test.dir/core/plan_executor_test.cc.o -MF caffe2/CMakeFiles/plan_executor_test.dir/core/plan_executor_test.cc.o.d -o caffe2/CMakeFiles/plan_executor_test.dir/core/plan_executor_test.cc.o -c ../caffe2/core/plan_executor_test.cc 
Jan 08 14:21:33 ../caffe2/core/plan_executor_test.cc:110:17: error: no matching member function for call to 'push_back' 
Jan 08 14:21:33       args_ptrs.push_back(arg.data()); 
Jan 08 14:21:33       ~~~~~~~~~~^~~~~~~~~ 
Jan 08 14:21:33 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/vector:677:36: note: candidate function not viable: 1st argument ('const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::value_type *' (aka 'const char *')) would lose const qualifier 
Jan 08 14:21:33     _LIBCPP_INLINE_VISIBILITY void push_back(const_reference __x); 
Jan 08 14:21:33                                    ^ 
Jan 08 14:21:33 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/vector:680:36: note: candidate function not viable: 1st argument ('const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::value_type *' (aka 'const char *')) would lose const qualifier 
Jan 08 14:21:33     _LIBCPP_INLINE_VISIBILITY void push_back(value_type&& __x); 
Jan 08 14:21:33                                    ^ 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

d4l3k committed:
caffe2/core/plan_executor: add cancellation of async nets on error + propagate exceptions via std::exception_ptr for stack traces (#31966)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31966

This has three parts:
* When `--caffe2_handle_executor_threads_exceptions` is set when a parallel execution step throws an exception it can hang waiting for async nets to finish. This adds cancellation code to cancel any async nets.
* This makes the exceptions returned from parallel workers pass a std::exception_ptr so the stack trace can be recorded with folly::SmartExceptionTracer.
* Define Cancel method at NetBase level to avoid pulling in unsupported AsyncSchedulingNet for fbandroid.

Test Plan:
Added unit tests for plan_executor

  buck test //caffe2/caffe2:caffe2_test_cpu
  buck test //caffe2/caffe2:caffe2_test_cpu -- PlanExecutorTest --stress-runs 100

Reviewed By: boryiingsu

Differential Revision: D19320177

fbshipit-source-id: 2739c51bad1aea011a598da48df2f18866eb6926
